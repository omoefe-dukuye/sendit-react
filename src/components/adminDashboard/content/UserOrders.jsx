import React, { Component } from 'react';
import '@babel/polyfill';
import { toast } from 'react-toastify';
import axios from '../../../utils/axiosConfig';
import Loader from '../../loader';
import AllParcels from './allParcels';

export class CancelOrder extends Component {
  state = {
    userId: '',
    ParcelIdErrorMessage: ''
  };

  onParcelIdChange = ({ target: { value: userId } }) => {
    this.setState({
      userId,
      userIdErrorMessage: ''
    });
  };

  onParcelIdBlur = async () => {
    const { userId } = this.state;
    if (!Number(userId)) {
      return this.setState({ userIdErrorMessage: 'must be a number.' });
    }
  }


  handleSubmit = async (e) => {
    e.preventDefault();
    const { userId } = this.state;

    if (!Number(userId)) {
      return this.setState({ userIdErrorMessage: 'must be a number.' });
    }

    this.setState({ buttonDisabled: true, loading: true });

    try {
      const { data: { orders: parcels } } = await axios.get(`${process.env.API_ROOT_URL}/users/${userId}/parcels`);
      this.setState({ parcels });
    } catch ({ response: { data: { error } } }) {
      toast.error(error);
    } 

    this.setState({ buttonDisabled: false, loading: false });
  }

  render() {
    const {
      state: {
        buttonDisabled,
        loading,
        userId,
        parcels = [],
        userIdErrorMessage,
      },
      onParcelIdBlur,
      onParcelIdChange,
      handleSubmit,
    } = this;

    const message = userIdErrorMessage || false;


    const disableButton = buttonDisabled || message;


    return (
      <div className='form__page create-order user-orders'>
        <div className='form__container'>
          <form className='form create-order__form'>
            <div className='form__input-top'>
              <p>
                <span className='form__input-top__label'>User ID</span>
                <span className='form__input-top__error'>{userIdErrorMessage}</span>
              </p>
            </div>
            <input
              className={userIdErrorMessage ? 'red-border' : ''}
              type='number'
              name='userId'
              value={userId}
              onChange={onParcelIdChange}
              onBlur={onParcelIdBlur}
              required
            />
            <button
              disabled={disableButton}
              className='button create-order__form__button'
              onClick={handleSubmit}
            >
              <span className={`button__text--create-order ${buttonDisabled ? 'hidden' : undefined}`}>
                Get Orders
              </span>
              <span className={buttonDisabled ? 'spinner spinner--button' : undefined}></span>
            </button>
          </form>
            {
              loading
                ? <Loader />
                : <AllParcels parcels={parcels} />
            }
          </div>
      </div>
    );
  }
}

export default CancelOrder;
