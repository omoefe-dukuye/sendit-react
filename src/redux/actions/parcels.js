import { toast } from "react-toastify";
import axios from "../../utils/axiosConfig";
import { ADD_ORDER } from "./types";

const api = process.env.API_ROOT_URL;

export const createOrder = (payload) => async (dispatch) => {
  try {
    const { data: { parcel } } = await axios.post(`${api}/parcels`, payload);

    dispatch({
      type: ADD_ORDER,
      payload: parcel
    });

    return parcel;
  } catch ({ response: { data: { error } } }) {
   toast.error(error);
  }
}

export default createOrder;
