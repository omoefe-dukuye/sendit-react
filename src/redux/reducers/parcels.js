import { ADD_ORDER } from '../actions/types';

const parcelReducerDefaultState = [];

export default (state = parcelReducerDefaultState, action) => {
  switch(action.type) {
    case ADD_ORDER:
      return [...state, action.payload];

    default:
      return state;
  }
};
