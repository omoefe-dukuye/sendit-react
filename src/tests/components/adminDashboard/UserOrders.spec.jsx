import React from 'react';
import { shallow } from 'enzyme';
import '@babel/polyfill';
import axios from '../../../utils/axiosConfig';
import parcels from '../../mocks/mockParcels';
import UserOrders from '../../../components/adminDashboard/content/UserOrders';

jest.mock('../../../utils/axiosConfig');

const tick = () => {
  return new Promise(resolve => {
    setTimeout(resolve, 0);
  })
};

test('Should render static component', async () => {
  const wrapper = await shallow(<UserOrders />);
  expect(wrapper).toMatchSnapshot();
});

test('Should handle input interactions', async () => {
  const wrapper = await shallow(<UserOrders />);
  wrapper.find('input').simulate('change', { target: { value: 3 } });
  expect(wrapper.state('userId')).toBe(3);

  wrapper.find('input').simulate('blur');
  expect(wrapper.state('userIdErrorMessage')).toBe('');

  wrapper.find('input').simulate('change', { target: { value: 'p' } });
  expect(wrapper.state('userId')).toBe('p');

  wrapper.find('input').simulate('blur');
  expect(wrapper.state('userIdErrorMessage')).toBe('must be a number.');
});

test('Should handle submit interactions', async () => {
  axios.get.mockRejectedValue({ response: { data: { error: 'error' } } });
  const wrapper = await shallow(<UserOrders />);
  wrapper.find('button').simulate('click', { preventDefault: jest.fn() });
  expect(wrapper.state('userIdErrorMessage')).toBe('must be a number.');
  
  wrapper.find('input').simulate('change', { target: { value: 2 } });
  expect(wrapper.state('userId')).toBe(2);
  
  wrapper.find('button').simulate('click', { preventDefault: jest.fn() });

  axios.get.mockResolvedValue({ data: { orders: parcels } });
  wrapper.find('button').simulate('click', { preventDefault: jest.fn() });
  await tick();
  expect(wrapper.state('parcels')).toEqual(parcels);

});



