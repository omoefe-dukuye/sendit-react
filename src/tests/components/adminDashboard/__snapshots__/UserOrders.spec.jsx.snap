// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render static component 1`] = `
<div
  className="form__page create-order user-orders"
>
  <div
    className="form__container"
  >
    <form
      className="form create-order__form"
    >
      <div
        className="form__input-top"
      >
        <p>
          <span
            className="form__input-top__label"
          >
            User ID
          </span>
          <span
            className="form__input-top__error"
          />
        </p>
      </div>
      <input
        className=""
        name="userId"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        type="number"
        value=""
      />
      <button
        className="button create-order__form__button"
        disabled={false}
        onClick={[Function]}
      >
        <span
          className="button__text--create-order undefined"
        >
          Get Orders
        </span>
        <span />
      </button>
    </form>
    <AllParcels
      parcels={Array []}
    />
  </div>
</div>
`;
